### ISSUE TITLE
Massive amount of CPU cycles in use

### ISSUE THREAD

**Issue by pbrooker** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#3332]
**Checklist:**

* [ ] I have checked that there are no issues with similar or the same content

  **YOU SHOULD CHECK CLOSED ISSUES ASWELL**
* [ ] I have checked the FAQ (https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/wiki/FAQ) and the answer I am looking for is not there
* [ ] I have double checked and can reproduce the issue

**OS:**
Mac High Sierra v 10.13.6
**GPMDP Version:**
4,6,1
**Issue Descriptions:**
Latest version is using an insane amount of CPU cycles, causing everything else to run slow when in use.
**Steps to Reproduce:**

Run the software
![screen shot 2018-08-21 at 9 13 46 am](https://user-images.githubusercontent.com/14824835/44400710-91e56080-a522-11e8-91ab-9655b9e13f75.png)

**Comment by diveint** [comment]
Experiencing this exact same thing.

**Comment by bbrcan** [comment]
Can confirm I am also experiencing this issue on Ubuntu 18.04. Have had to stop using the app

**Comment by alexgred** [comment]
High CPU usage on ubuntu 18.04. Average usage 30%. i7 7700.

**Comment by Deckee** [comment]
Ubuntu 18.04, 20-50% cpu on my 3840qm and 10% gpu amd M4000 latest stable

**Comment by mintiefresh** [comment]
Experiencing this exact same issue as well.

macOS High Sierra 10.13.6
GDMP 4.6.1

**Comment by Weafyr** [comment]
Same on Linux Mint 19 (64bit), constant usage of 16% CPU
i7-2640M
GDMP 4.6.1

**Comment by oms064** [comment]
Same issue here with macOs HighSierra 10.13.6. Have to wait a little while before noticing the lag.

**Comment by theLastOfCats** [comment]
Confirming.

OS: High Sierra 10.13.6
GPMDP version: 4.6.1

**Comment by j3g** [comment]
CPU usage hovers around 30%  (OSX 10.11.6)

**Comment by alexriabtsev** [comment]
Confirming. MacOS Mojave 10.14

**Comment by maxim-vorotilo** [comment]
confirm. mac os mojave 10.14

**Comment by djfjeff** [comment]
I have the same high CPU and do not have Discord Rich Presence enabled.

Fedora 28

**Comment by felipemlemos** [comment]
Confirming. macOS 10.14.
I never used  Discord Rich Presence.
Just after opened the application without any song playing the CPU usage is around 12% to 15%.

**Comment by Minothor** [comment]
Same issue on Pop_OS (essentially Ubuntu 18 LTS)
High CPU and (thankfully stable at 2.1 GB) RAM usage, mitigated it by setting process priority to low,  but it's still someone noticeable in how rapidly it responds to control input (even before lowering priority)

**Comment by bmoyles** [comment]
Similar problems here on Ubuntu 18.04. If I minimize the app to the tray, CPU is reasonable, but if I have the player showing at all, even in the mini player, I end up with  two processes that hover around 15-20% CPU utilization.

**Comment by Dan33185** [comment]
Just installed this on Windows 7 and had to uninstall within about 15 minutes, almost unusable because of the amount of resources it uses.

**Comment by Dan33185** [comment]
Any response from the Dev yet?

**Comment by kazak1377** [comment]
Same issue. On macOS mojave over 100% CPU ussage and over 6GB virtual memory ussage

**Comment by AlexMStudioLabs** [comment]
Can also confirm. macOS 10.14.1 and GPMPD 4.6.1. The program is borderline unusable in its current state. Would love to see this resolved.

**Comment by marstj** [comment]
I've had this for some time, and it's in the latest version 4.6.1 too. It starts right away, even before playing the first song. Fresh Debian sid on x86_64.

All I can see is that it's one of the two "Google Play Mus" threads that's busy looping. Attaching ~10s of ["strace -r" log](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/files/2626957/strace.log) FWIW, but it just shows futexes being woken and not waited upon.

**Comment by kazak1377** [comment]
By the way problem is not in Discord integration. I've turned it off and app stills producing several processes in total amount of 100% CPU ussage

**Comment by wmertens** [comment]
It used to take quite a bit of CPU in Chrome as well but in v70 Google Play Music is only taking 2-6% when the window is minimized.

The Desktop app is nicer, but I can't afford the extra CPU so I made this applescript to toggle the player window, and I trigger it with Quicksilver:

`Toggle Google Play Music.scpt`:
```applescript
tell application "Google Chrome"
	if frontmost then
		set w to window 1
		if visible of w is true and (URL of active tab of w) contains "music/listen" then
			-- Chrome uses minimized instead of miniaturized
			set minimized of w to true
			return
		end if
	end if
	activate
	set i to 0
	repeat with w in (windows)
		set i to i + 1
		set j to 0
		repeat with t in (tabs of w)
			set j to j + 1
			if URL of t contains "music/listen" then
				set (active tab index of w) to j
				set index of w to 1
				tell application "System Events" to tell process "Google Chrome"
					perform action "AXRaise" of window 1 -- `set index` doesn't always raise the window
					-- keystroke "f" using {option down, command down}
				end tell
				return
			end if
		end repeat
	end repeat
	activate
	open location "https://play.google.com/music/listen"
end tell
```

**Comment by chuckmitchell** [comment]
Can confirm on macOS Mojave 10.14.2 I see 3 processes with ~100% cpu, ~13%, %5 respectively.
Player is minimized.

**Comment by px4n** [comment]
Can confirm this on Fedora 29 as well.

```
LSB Version:	:core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
Distributor ID:	Fedora
Description:	Fedora release 29 (Twenty Nine)
Release:	29
Codename:	TwentyNine
```

```
Linux 4.19.5-300.fc29.x86_64 #1 SMP Tue Nov 27 19:29:23 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

CPU usage is unbearable and grinds machine to a halt on a modern quad-core laptop, even when it's not playing any music, so I've had to uninstall.

Will be watching this threads in the hope that it goes fixed soon, as it is such a good application.

**Comment by Dan33185** [comment]
This seems to be a dead project, there hasn't been a response by the developer or a new release in 5 months.

**Comment by px4n** [comment]
His Github is active, I wonder if reaching out to him to understand current status is in order.

@MarshallOfSound sorry to highlight was wondering on the status of this project?
Let us know if we need to open a new issue to discuss.

**Comment by px4n** [comment]
Hi @jostrander ,

Thanks for the swift reply and quick digging that you've done around the issue, it is appreciated!

I believe most understand that this is a side-project (as with most OSS projects out there), and are very grateful for this (I know I am) :100: 

I believe clarification is wanted on whether or not the project is dead or needs developer assistance.

If that is made clear, I'm pretty sure there will be others that will come forward to assist and/or take-over as needed.

I see that you are a collaborator, so am I right in assuming that your response means the project is not dead however could do with some help from the community?

**Comment by jostrander** [comment]
Community help is always appreciated, and no, I wouldn't consider it a dead project :+1:

**Comment by px4n** [comment]
Thanks for clarifying! 

I might start looking under the hood as I get time (unfortunately JavaScript is not my first language), but let me know if there is anything I can do to assist (testing, etc.).

**Comment by MarshallOfSound** [comment]
@px4n @Dan33185 I feel like I should clarify a few things here.

1. This project is **definitely not dead** and something to be aware of, I find it quite insulting when one of my projects is called dead.  A project is declared dead / abandoned by its maintainers, not onlookers in GitHub issues.

On a brighter note:

2. I don't have a significant amount of time nowadays to fix bugs / do things with this player.  If some community members want to step up to the plate and start landing stuff in PRs I'm 💯% on board with that and as with all my projects will gladly carve time out of my day to review those pull requests.

If any folks have questions about the layout of the code, ask away, if anyone wants to refactor half the app, go for it.  Please be aware that half the code here at least was written like 3+ years ago 😆 .

TLDR: Come help all ye' who want to help 😄

**Comment by MarshallOfSound** [comment]
@px4n I'm not that active in Gitter nowadays so you're best bet tbh is to do the old fashioned thing and email me.  Also if you put up just WIP PR's and ask for clarification on those I can respond there.

**Comment by inkubux** [comment]
I don't know what they do differently , but I found a simple electron google play app that uses 3%-7% instead of 30% on my macbook.

https://github.com/pmsaue0/play

The app is not configurable as this one but it's an alternative for now

**Comment by kazak1377** [comment]
@inkubux thank for your suggestion. But this app cant be controlled via media keys and dont support now playing. Waiting for fix GPMDP.

**Comment by bmoyles** [comment]
I checked out master and applied @jostrander 's fixes in #3441 but they didn't seem to help much.

I opened dev tools, closed the main window, and started the CPU profiler. I let it run for ~ 25s, clicked a bit in the tray menu, and then restored the main window.  For the last ~5-10s I had the main window visible and moved my mouse around a bit in the window.

Gzipped profile is attached, looks to be some React foo (my JS is *weak* at best). Once the main window is visible, a batchedUpdates call seems to fire every 20ms or so, occasionally handling SyntheticMouseEvents
It's immediately obvious from the flame graph view when I restore the main window, should make it easy to drill down if you should load the profile yourself.
[bmoyles-gpmdp.cpuprofile.gz](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/files/2700867/bmoyles-gpmdp.cpuprofile.gz)

**Comment by bmoyles** [comment]
Did a bit more digging, and near as I can tell, despite #3441 hiding the loader, it continues to animate, so I stopped the animation via the console. Better, but I noticed that its very presence of the loader SVG was still causing layout recalculations. So, again via the console, I removed the SVG and CPU seems normal again

**Comment by jostrander** [comment]
That's completely different from the profiling I get currently. Do you have lyric scrolling enabled?

**Comment by bmoyles** [comment]
@jostrander I take it back, I applied the fix from #3441 but didn't have yarn watch running or do a yarn build. Sigh. 

So #3441 definitely seems to make a significant improvement, SVG isn't causing any problems with that fix in place.

There is still a decent spike in CPU if the main window has focus and the mouse cursor is moving, though.  Attached is a fresh profile from the Performance tab of the dev tools.  Roughly 5 seconds with window un-focused, then I switch to the main window and move my mouse around in an empty section for ~10s, and then I remove focus from the main window for the remaining 5s.  While mouse is moving, react appears to dispatch every 20ms, followed by painting/compositing/etc.  While this is happening, the two render processes and the main process jump to 20-35% CPU. Whether or not this behavior existed prior to 4.6.1 I can't say.

**Comment by bmoyles** [comment]
[bmoyles-gpmdp-profile.json.gz](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/files/2701234/bmoyles-gpmdp-profile.json.gz)
Forgot attachment.

**Comment by aldevel** [comment]
Although I use it through Flatpak's flathub/com.googleplaymusicdesktopplayer.GPMDP on KDE neon, it's  still issue on my system too.
https://github.com/flathub/com.googleplaymusicdesktopplayer.GPMDP

**Comment by kazak1377** [comment]
4.6.1 still no fix

**Comment by theLMGN** [comment]
@kazak1377 gdpmp hasn't been updated since this issue was opened 🙄

**Comment by dmillerzx** [comment]
@bmoyles while I don't have any evidence on hand, I do not think this behavior is limited to 4.6.1, I've notice what seems like 'extreme' resources usage, since I started using this app almost two years ago.

**Comment by chrisspiegl** [comment]
Is there a plan on updating GDPMP in the near future? I love how it works but specifically this issue is a pretty big problem fo rm.

**Comment by tankerkiller125** [comment]
I notice that out of 4 cores on my CPU one of them is always at 100% when this application is open, plus its using 2GB of ram! Its way to much ram to be reasonable.

**Comment by OranguTech** [comment]
+1 Here, Windows 10. GPMDP 4.6.1 (installed via Chocolately.) With main window open (On home or library, not playing or anything,) doesn't have to be active, bounces between 1-8% CPU. (Meanwhile, Chrome w/ 37 tabs is c. 1.5%). Minimize window, drops to 0.

**Comment by MarshallOfSound** [comment]
Gonna close this out as significantly improved by #3441
